### PR TITLE
KHP3-277 Added Months of prescription in Active on ART Linelist

### PR DESCRIPTION
**Pull request by CynthiaKamau**
Month Of Prescription

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@Handler(supports= ETLMonthsOfPrescriptionDataDefinition.class, order=50)
+public class ETLMonthsOfPrescriptionDataEvaluator implements PersonDataEvaluator {
+
+    @Autowired
+    private EvaluationService evaluationService;
+
+    public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationContext context) throws EvaluationException {
+        EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
+
+        String qry="select patient_id,\n" +
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:37]: I thought this query should basically compute the time difference in months between the last visit date and the appointment date given on that particular visit. You can use timestampdiff to simplify the computation

File: api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                "CASE \n" +
+                "WHEN refill_date is null then \n" +
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:39]: I don't think we need to use the refill date. @patryllus do you have some insight?

**Review by ojwanganto — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java (hunk 2 above)
Comment (CynthiaKamau) [reply api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:39]: What date should i use?

Comment (patryllus) [reply api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:39]: Last appointments date (TCA date)

File: api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Autowired
+    private EvaluationService evaluationService;
+
+    public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationContext context) throws EvaluationException {
+        EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
+
+        String qry="SELECT patient_id, TIMESTAMPDIFF(MONTH, date(max(visit_date)), date(mid(max(concat(visit_date,next_appointment_date )),11))) as months_of_prescription\n" +
+                "FROM kenyaemr_etl.etl_patient_hiv_followup\n" +
+                "WHERE (select date(mid(max(concat(visit_date,next_appointment_date )),11))) is not null\n" +
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:39]: This line is a repetition of line 37. I think the WHERE clause should just read 'where next_appointment_date is not null. Is there a need to also consider when a patient had a physical visit? @makombe @njorocs @patryllus

**Review by ojwanganto — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java (hunk 3 above)
Comment (njorocs) [reply api/src/main/java/org/openmrs/module/kenyaemr/reporting/data/converter/definition/evaluator/art/ETLMonthsOfPrescriptionDataEvaluator.java:39]: I don't think we need to limit the results to visits where the patient was present

**Review by ojwanganto — APPROVED** [review]

**Pull request merged**